ts/tree/dbTree.ts

```diff
@@ -1,17 +1,218 @@
-import { Hashable } from "../interfaces";
-import { Hasher } from "./hasher";
-import { Leaf, LeafFactoryFunc } from "./leaves/Leaf";
-import { MemoryTree } from "./memoryTree";
-
-export class DBTree<LeafType extends Leaf<Hashable>> extends MemoryTree {
-    private readonly leafFactory: LeafFactoryFunc<LeafType>;
-
-    constructor(
-        depth: number,
-        leafFactory: LeafFactoryFunc<LeafType>,
-        hasher?: Hasher
-    ) {
-        super(depth, hasher || Hasher.new());
-        this.leafFactory = leafFactory;
+import {
+    BadMergeAlignment,
+    EmptyArray,
+    ExceedTreeSize,
+    MismatchHash,
+    MismatchLength,
+    NegativeIndex
+} from "../exceptions";
+import { Hasher, Node } from "./hasher";
+import { ItemNode } from "./leaves/Node";
+import { AsyncTree, Data, Witness } from "./tree";
+
+export class DBTree implements AsyncTree {
+    public readonly zeros: Array<Node>;
+    public readonly depth: number;
+    public readonly setSize: number;
+    public readonly hasher: Hasher;
+    public readonly nodeType: string;
+    public _root: Node;
+
+    public static new(depth: number, nodeType: string, hasher?: Hasher) {
+        return new DBTree(depth, nodeType, hasher || Hasher.new());
+    }
+
+    constructor(depth: number, nodeType: string, hasher: Hasher) {
+        this.depth = depth;
+        this.hasher = hasher;
+        this.zeros = this.hasher.zeros(depth);
+        this.setSize = 2 ** this.depth;
+        this.nodeType = nodeType;
+        this._root = this.zeros[0];
+    }
+
+    get root(): Node {
+        return this._root;
+    }
+
+    public async getNode(level: number, index: number): Promise<Node> {
+        try {
+            return await ItemNode.fromDB(this.nodeType, level, index);
+        } catch (e) {
+            return this.zeros[level];
+        }
```

Comment:
This could suppress other errors asides from the item not being found, can we do an undefined check on `ItemNode.fromDB` and throw instead?

Reply:
check if key exists